README
Readme for the ouplayer (ouembed) project.


* http://embed.open.ac.uk | http://media-podcast-dev.open.ac.uk/player
* Bugs/ Issues:  http://iet-it-bugs.open.ac.uk/project/issues/ouplayer


OU player: Copyright © 2010-2011 The Open University. All rights reserved.
 * Not licensed as open-source (yet!)
 * Author: Nick Freear <n.d.freear+@+open.ac.uk> / Institute of Educational Technology.

OUICE-Dark theme: Copyright 2011 The Open University. All rights reserved.
 * Author: David Winter <d.j.winter+@+open.ac.uk> / LTS.

OUICE-Bold theme: Copyright 2011 The Open University. All rights reserved.
 * Author: Anthony McEvoy <a.mcevoy+@+open.ac.uk> / IT.

OUICE-Light theme: Copyright 2011 The Open University. All rights reserved.
 * Author: David Winter <d.j.winter+@+open.ac.uk> / LTS.

OU player/OU embed translations.
 * Kiswahili/ Swahili: Pauline Nngimwa, Christian Msoffe.
 * Chinese Simplified: Guangyu Yang Freear.

Based on codeigniter-reactor, changeset 2263:1b173956c908 (v2.0.0+) (6 mar.-10 days)
 *  <http://bitbucket.org/ellislab/codeigniter-reactor/changeset/1b173956c908>

CodeIgniter: Copyright (c) 2008 - 2011, EllisLab, Inc.
 * License: <http://codeigniter.com/user_guide/license.html>

jquery-oembed: Copyright Richard Chamorro, Avanzis.
 * MIT License: <http://opensource.org/licenses/MIT>
 * Code: <http://code.google.com/p/jquery-oembed/#r20>

Flowplayer: Copyright (c) 2008 Flowplayer Ltd.
 * GPLv3 License: <http://gnu.org/licenses/gpl.html>
 * Code: <http://flowplayer.org/>

Flowplayer Controlbar Javascript plugin: (c) 2008 Flowplayer Ltd. / Modifications ©2011 The Open University.
 * Dual licensed under MIT and GPL 2+ licenses.
 * Code: <http://flowplayer.org/plugins/javascript/controlbar.html>

Fixed Tooltip script (ouplayer.tooltips.js): © Dynamic Drive (www.dynamicdrive.com) / Modifications ©2011 The Open University.
 * License: <http://dynamicdrive.com/notice.htm>
 * Code: <http://dynamicdrive.com/dynamicindex5/fixedtooltip.htm>

php-po-parser: Copyright (C) 2008, Iulian Ilea (http://iulian.net), all rights reserved.
 * License: <http://gnu.org/licenses/gpl.html> GNU GPL v3 onwards (code), or GNU Lesser GPL (project page)
 * Code: <http://code.google.com/p/php-po-parser/>

Fire PHP: Copyright (c) 2006-2009, Christoph Dorn.
 * License: <http://opensource.org/licenses/BSD-3-Clause> New BSD
 * Code: <http://www.firephp.org/>



TODOs:
 * Add INSERTs to *_serv.php
 * Use $config[providers]..examples in the Demo controller.
 * Error handling - use exceptions.
 * Add $thumbnail_width to embed_cache Database.
 * services.json, like Embed.ly.
 * Use $this->CI->oembed_request, etc.


[End.]


docs/CHANGELOG.md
# Open Media Player

## Version history

Versions 1.2+ and 2.x. Please see:

* <https://github.com/IET-OU/open-media-player/releases>

### v1.1-26-g731260f
Release: 5 Jan/ Live: 14 Jan 2015

* Fix jQuery XSS vulnerability [Bug:8]

### v1.1-25-g535a7b7
Release: 17 Dec/ Live: 18 Dec 2014

* Fix PHP notice/ warning [Bug:7]
* Fix PHP `error_reporting()`/ `display_errors` use [Bug:7]

### v1.1-21-gba86785
Release: 5 Dec/ Live: 11 Dec 2014

* Fix to add "no-proxy" support to Http library - intranet-restricted [Bug:3]
* Fix cookies for intranet-restricted transcripts [Bug:4]
* Add analytics events to OpenLearn (legacy) player [Bug:5]

### v1.1-12-g1f75bbf
Release: 21 Nov/ Live: approx. 26 Nov 2014

* Fix OU Media Player for intranet-restricted content [Bug:1]

### v1.1-9-g17ff3ce
<!-- v1.1-0-g0075a19 equiv. to v1.0-rc.1-45-g0075a19 // v1.1-8-g7cf9e8f -->
Release: ~2 Jun/ Live: 12 Jun 2014

* Hide the title panel/ banner for OUICE/OU-branded pages [iet-it-bugs: 1486](http://iet-it-bugs.open.ac.uk/node/1486) [LTS-redmine: 10744](http://ltsredmine.open.ac.uk/issues/10744),
* Fix: "In IE11 (Win8) we just see a black rectangle.." [iet-it-bugs: 1487](http://iet-it-bugs.open.ac.uk/node/1487) (CSS)
* Fix: broken transcript links,
* Fix: Remove "View on Podcast site" link for private media,
* Fix: Correctly handle secure HTTPS media URLs,
* Fix: add support for "http://media-podcast.open.ac.uk" URLs,
* Fix: remove "role=application" from Player embeds (accessibility),
* Fix: time-display widget overlaps progress-bar - prevents dragging (CSS),
* Add Google Analytics to OpenLearn player variant, add Player logo to about page.

### v1.0-rc.1-32-gd251c7a
Release: ~10 Dec/ Live: 12 Dec 2013

* Fix: full screen button, only top half works - FLV [Ltsredmine: #8526][lts#:8526]
* Fix: "no-Flash" message intermittently appearing (OpenLearn) [#9072]
* Fix: I'm seeing subtitles twice - Chrome only [#9071]
* Fix: full screen consistency, inc. follow up [Ltsredmine: #7911][lts#:7911]
* Fix: Player controls alignment CSS [iet-it-bugs: #1485][#:1485]

### v1.0-rc.1-0-gd0bb0cb
v0.95-beta-80-gd0bb0cb
Release: 25 July/ Live: approx. 30 July 2013

#### Player fixes following [LTS-tech-testing][lts-bugs]:

* List - todo - [iet-it-bugs: #1477][#:1477]
* Video offset bug [Ltsredmine: #6932][lts#:6932]
* Playing on tablet device [Ltsredmine: #7182][lts#:7182]
* ..

### v0.95-beta-39-g584d305
Release: 31 May/ Live: approx. 21 June 2013

### v0.95-beta-30-gcfe9c2d
Release: 20 May/ Live: approx. 23 May 2013

#### OU Player features:

* The Player is now available under HTTPS/ SSL - initially for VLE use.

#### OU Player bug fixes:

* Javascript/ configuration fix for video size - Internet Explorers/ MSIEs 7/ 9 [Bug #1474][#:1474]
* Config. fix to Ender/jeesh Javascript URL under HTTPS/ SSL [Bug #1473]
* Fix for `no-svg` class in `oup-light` CSS stylesheet [Bug #1476]
* Follow-up PHP fix for analytics [Bug #1464]

### v0.95-beta-2-gb72490b
Release: 31 January/ Live: approx. 7 February 2013

#### OU Player features:

* Finished support for "rgb" colour parameter - VLE player [Bug #1324][#:1324]; fix 2013-01-07,
* Use new HTML transcript from OU Podcast site [Bug #1460][#:1460]; fix 2013-01-22; ref #1409],

#### OU Player bug fixes:

* Fixed video size issue - Chrome etc. - ender/jeesh/VLE/ no captions [Bug #1456][#:1456]; fix 2013-01-11; reported by Ray.Guo,
* Tooltip styling/ Z-index bug [Bug #1458]
* Podcast player "private" flag bug [Bug #1448; fix 2013-01-10; reported by Ben.Hawkridge]
* Firefox VLE player bug - "Sorry, your browser appears.." [Bug #1457][#:1457]; fix 2013-01-25; reported by Ray.Guo; [Bug #1447][#:1447]; reported by Ben.Hawkridge,
* "rgb" parameter error handling too draconian [Bug #1453; fix 2013-01-07]
* Fixed access control - ignore private/ published flags, set cookie [Bug #1463][#:1463]; fix 2013-01-30; reported by Ben.Hawkridge,

#### OU Embed:

* OU Embed demo/preview form [Bug #1455][#:1455]
* Bibsonomy provider, ScraperWiki external provider [Bug #1461; #1420]
* Embedding from Wordpress.com blogs - Noembed [Cloudworks #310]
* Fileviewer provider, including CompendiumLD SVG [Bug #1420]
* iSpot external provider [Bug #1408; added 2012-10-11]

### v0.9-beta-123-gfc4eaf3
Release: 10 December/ Live: 13 December 2012

* Upgrade to latest MediaElement.js 2.10.0 [Bug #1368][#:1368]; commit 2012-12-10; 2.9.5-32-g98263df,
* Upgrade to CodeIgniter 2.1.3 [Bug #1410; commit 2012-10-31]
* Support "maxwidth" - plus experimental "pcwidth" (100% / percent width) [Bug #1415]
* New placeholder/locked image for private podcasts [Bug #1401; added 2012-09-25; by Peter Devine]

#### OU Player bug fixes (September-December 2012):

* Chrome fails MP4 using mediaelement/ maybe-to-no [Bug #1416][#:1416]; fix 2012-12-10,
* PHP-Apache hangs on IT-EUD-Acct/ LOCK_EX/ CodeIgniter logging [Bug #1446; fix 2012-12-10]
* MSIE 9 compatibility view style/CSS width [Bug #1417]
* Mobile fix - width=100%/ jQuery/Ender [Bug #1414]
* Documentation [Bug #1413]
* MY_Input workaround for Drupal-oembed consumers [Bug #1378; fixed 2012-11-06]
* MediaElement.js error/ event handling [Bug #1412; fixed 2012-10-29]
* IT deployment fixes [Bug #1406][#:1406]; #1400; September,


[bugs]: http://iet-it-bugs.open.ac.uk/project/issues/ouplayer
[lts-bugs]: http://ltsredmine.open.ac.uk/projects/ouplayer/issues
[#:*]: http://iet-it-bugs.open.ac.uk/node/
[lts#:*]: http://ltsredmine.open.ac.uk/issues/
[bug#:*]: https://github.com/IET-OU/ouplayer/issues/
[Bug:*]:  https://github.com/IET-OU/ouplayer/issues/


system/core/Common.php
<?php  if ( ! defined('BASEPATH')) exit('No direct script access allowed');
/**
 * CodeIgniter
 *
 * An open source application development framework for PHP 5.1.6 or newer
 *
 * @package		CodeIgniter
 * @author		ExpressionEngine Dev Team
 * @copyright	Copyright (c) 2008 - 2011, EllisLab, Inc.
 * @license		http://codeigniter.com/user_guide/license.html
 * @link		http://codeigniter.com
 * @since		Version 1.0
 * @filesource
 */

// ------------------------------------------------------------------------

/**
 * Common Functions
 *
 * Loads the base classes and executes the request.
 *
 * @package		CodeIgniter
 * @subpackage	codeigniter
 * @category	Common Functions
 * @author		ExpressionEngine Dev Team
 * @link		http://codeigniter.com/user_guide/
 */

// ------------------------------------------------------------------------

/**
* Determines if the current version of PHP is greater then the supplied value
*
* Since there are a few places where we conditionally test for PHP > 5
* we'll set a static variable.
*
* @access	public
* @param	string
* @return	bool	TRUE if the current version is $version or higher
*/
if ( ! function_exists('is_php'))
{
	function is_php($version = '5.0.0')
	{
		static $_is_php;
		$version = (string)$version;

		if ( ! isset($_is_php[$version]))
		{
			$_is_php[$version] = (version_compare(PHP_VERSION, $version) < 0) ? FALSE : TRUE;
		}

		return $_is_php[$version];
	}
}

// ------------------------------------------------------------------------

/**
 * Tests for file writability
 *
 * is_writable() returns TRUE on Windows servers when you really can't write to
 * the file, based on the read-only attribute.  is_writable() is also unreliable
 * on Unix servers if safe_mode is on.
 *
 * @access	private
 * @return	void
 */
if ( ! function_exists('is_really_writable'))
{
	function is_really_writable($file)
	{
		// If we're on a Unix server with safe_mode off we call is_writable
		if (DIRECTORY_SEPARATOR == '/' AND @ini_get("safe_mode") == FALSE)
		{
			return is_writable($file);
		}

		// For windows servers and safe_mode "on" installations we'll actually
		// write a file then read it.  Bah...
		if (is_dir($file))
		{
			$file = rtrim($file, '/').'/'.md5(mt_rand(1,100).mt_rand(1,100));

			if (($fp = @fopen($file, FOPEN_WRITE_CREATE)) === FALSE)
			{
				return FALSE;
			}

			fclose($fp);
			@chmod($file, DIR_WRITE_MODE);
			@unlink($file);
			return TRUE;
		}
		elseif ( ! is_file($file) OR ($fp = @fopen($file, FOPEN_WRITE_CREATE)) === FALSE)
		{
			return FALSE;
		}

		fclose($fp);
		return TRUE;
	}
}

// ------------------------------------------------------------------------

/**
* Class registry
*
* This function acts as a singleton.  If the requested class does not
* exist it is instantiated and set to a static variable.  If it has
* previously been instantiated the variable is returned.
*
* @access	public
* @param	string	the class name being requested
* @param	string	the directory where the class should be found
* @param	string	the class name prefix
* @return	object
*/
if ( ! function_exists('load_class'))
{
	function &load_class($class, $directory = 'libraries', $prefix = 'CI_')
	{
		static $_classes = array();

		// Does the class exist?  If so, we're done...
		if (isset($_classes[$class]))
		{
			return $_classes[$class];
		}

		$name = FALSE;

		// Look for the class first in the local application/libraries folder
		// then in the native system/libraries folder
		foreach (array(APPPATH, BASEPATH) as $path)
		{
			if (file_exists($path.$directory.'/'.$class.'.php'))
			{
				$name = $prefix.$class;

				if (class_exists($name) === FALSE)
				{
					require($path.$directory.'/'.$class.'.php');
				}

				break;
			}
		}

		// Is the request a class extension?  If so we load it too
		if (file_exists(APPPATH.$directory.'/'.config_item('subclass_prefix').$class.'.php'))
		{
			$name = config_item('subclass_prefix').$class;

			if (class_exists($name) === FALSE)
			{
				require(APPPATH.$directory.'/'.config_item('subclass_prefix').$class.'.php');
			}
		}

		// Did we find the class?
		if ($name === FALSE)
		{
			// Note: We use exit() rather then show_error() in order to avoid a
			// self-referencing loop with the Excptions class
			exit('Unable to locate the specified class: '.$class.'.php');
		}

		// Keep track of what we just loaded
		is_loaded($class);

		$_classes[$class] = new $name();
		return $_classes[$class];
	}
}

// --------------------------------------------------------------------

/**
* Keeps track of which libraries have been loaded.  This function is
* called by the load_class() function above
*
* @access	public
* @return	array
*/
if ( ! function_exists('is_loaded'))
{
	function &is_loaded($class = '')
	{
		static $_is_loaded = array();

		if ($class != '')
		{
			$_is_loaded[strtolower($class)] = $class;
		}

		return $_is_loaded;
	}
}

// ------------------------------------------------------------------------

/**
* Loads the main config.php file
*
* This function lets us grab the config file even if the Config class
* hasn't been instantiated yet
*
* @access	private
* @return	array
*/
if ( ! function_exists('get_config'))
{
	function &get_config($replace = array())
	{
		static $_config;

		if (isset($_config))
		{
			return $_config[0];
		}

		// Is the config file in the environment folder?
		if ( ! defined('ENVIRONMENT') OR ! file_exists($file_path = APPPATH.'config/'.ENVIRONMENT.'/config.php'))
		{
			$file_path = APPPATH.'config/config.php';
		}

		// Fetch the config file
		if ( ! file_exists($file_path))
		{
			exit('The configuration file does not exist.');
		}

		require($file_path);

		// Does the $config array exist in the file?
		if ( ! isset($config) OR ! is_array($config))
		{
			exit('Your config file does not appear to be formatted correctly.');
		}

		// Are any values being dynamically replaced?
		if (count($replace) > 0)
		{
			foreach ($replace as $key => $val)
			{
				if (isset($config[$key]))
				{
					$config[$key] = $val;
				}
			}
		}

//ou-specific, PHP 5.6 compat [Bug: #37]
		//Was: return $_config[0] =& $config;
		$_config[0] =& $config;
		return $_config[0];
//ou-specific ends.
	}
}

// ------------------------------------------------------------------------

/**
* Returns the specified config item
*
* @access	public
* @return	mixed
*/
if ( ! function_exists('config_item'))
{
	function config_item($item)
	{
		static $_config_item = array();

		if ( ! isset($_config_item[$item]))
		{
			$config =& get_config();

			if ( ! isset($config[$item]))
			{
				return FALSE;
			}
			$_config_item[$item] = $config[$item];
		}

		return $_config_item[$item];
	}
}

// ------------------------------------------------------------------------

/**
* Error Handler
*
* This function lets us invoke the exception class and
* display errors using the standard error template located
* in application/errors/errors.php
* This function will send the error page directly to the
* browser and exit.
*
* @access	public
* @return	void
*/
if ( ! function_exists('show_error'))
{
	function show_error($message, $status_code = 500, $heading = 'An Error Was Encountered')
	{
		$_error =& load_class('Exceptions', 'core');
		echo $_error->show_error($heading, $message, 'error_general', $status_code);
		exit;
	}
}

// ------------------------------------------------------------------------

/**
* 404 Page Handler
*
* This function is similar to the show_error() function above
* However, instead of the standard error template it displays
* 404 errors.
*
* @access	public
* @return	void
*/
if ( ! function_exists('show_404'))
{
	function show_404($page = '', $log_error = TRUE)
	{
		$_error =& load_class('Exceptions', 'core');
		$_error->show_404($page, $log_error);
		exit;
	}
}

// ------------------------------------------------------------------------

/**
* Error Logging Interface
*
* We use this as a simple mechanism to access the logging
* class and send messages to be logged.
*
* @access	public
* @return	void
*/
if ( ! function_exists('log_message'))
{
	function log_message($level = 'error', $message, $php_error = FALSE)
	{
		static $_log;

		if (config_item('log_threshold') == 0)
		{
			return;
		}

		$_log =& load_class('Log');
		$_log->write_log($level, $message, $php_error);
	}
}

// ------------------------------------------------------------------------

/**
 * Set HTTP Status Header
 *
 * @access	public
 * @param	int		the status code
 * @param	string
 * @return	void
 */
if ( ! function_exists('set_status_header'))
{
	function set_status_header($code = 200, $text = '')
	{
		$stati = array(
							200	=> 'OK',
							201	=> 'Created',
							202	=> 'Accepted',
							203	=> 'Non-Authoritative Information',
							204	=> 'No Content',
							205	=> 'Reset Content',
							206	=> 'Partial Content',

							300	=> 'Multiple Choices',
							301	=> 'Moved Permanently',
							302	=> 'Found',
							304	=> 'Not Modified',
							305	=> 'Use Proxy',
							307	=> 'Temporary Redirect',

							400	=> 'Bad Request',
							401	=> 'Unauthorized',
							403	=> 'Forbidden',
							404	=> 'Not Found',
							405	=> 'Method Not Allowed',
							406	=> 'Not Acceptable',
							407	=> 'Proxy Authentication Required',
							408	=> 'Request Timeout',
							409	=> 'Conflict',
							410	=> 'Gone',
							411	=> 'Length Required',
							412	=> 'Precondition Failed',
							413	=> 'Request Entity Too Large',
							414	=> 'Request-URI Too Long',
							415	=> 'Unsupported Media Type',
							416	=> 'Requested Range Not Satisfiable',
							417	=> 'Expectation Failed',

							500	=> 'Internal Server Error',
							501	=> 'Not Implemented',
							502	=> 'Bad Gateway',
							503	=> 'Service Unavailable',
							504	=> 'Gateway Timeout',
							505	=> 'HTTP Version Not Supported'
						);

		if ($code == '' OR ! is_numeric($code))
		{
			show_error('Status codes must be numeric', 500);
		}

		if (isset($stati[$code]) AND $text == '')
		{
			$text = $stati[$code];
		}

		if ($text == '')
		{
			show_error('No status text available.  Please check your status code number or supply your own message text.', 500);
		}

		$server_protocol = (isset($_SERVER['SERVER_PROTOCOL'])) ? $_SERVER['SERVER_PROTOCOL'] : FALSE;

		if (substr(php_sapi_name(), 0, 3) == 'cgi')
		{
			header("Status: {$code} {$text}", TRUE);
		}
		elseif ($server_protocol == 'HTTP/1.1' OR $server_protocol == 'HTTP/1.0')
		{
			header($server_protocol." {$code} {$text}", TRUE, $code);
		}
		else
		{
			header("HTTP/1.1 {$code} {$text}", TRUE, $code);
		}
	}
}

// --------------------------------------------------------------------

/**
* Exception Handler
*
* This is the custom exception handler that is declaired at the top
* of Codeigniter.php.  The main reason we use this is to permit
* PHP errors to be logged in our own log files since the user may
* not have access to server logs. Since this function
* effectively intercepts PHP errors, however, we also need
* to display errors based on the current error_reporting level.
* We do that with the use of a PHP error template.
*
* @access	private
* @return	void
*/
if ( ! function_exists('_exception_handler'))
{
	function _exception_handler($severity, $message, $filepath, $line)
	{
		 // We don't bother with "strict" notices since they tend to fill up
		 // the log file with excess information that isn't normally very helpful.
		 // For example, if you are running PHP 5 and you use version 4 style
		 // class functions (without prefixes like "public", "private", etc.)
		 // you'll get notices telling you that these have been deprecated.
		if ($severity == E_STRICT)
		{
			return;
		}

		$_error =& load_class('Exceptions', 'core');

		// Should we display the error? We'll get the current error_reporting
		// level and add its bits with the severity bits to find out.
		if (($severity & error_reporting()) == $severity)
		{
			$_error->show_php_error($severity, $message, $filepath, $line);
		}

		// Should we log the error?  No?  We're done...
		if (config_item('log_threshold') == 0)
		{
			return;
		}

		$_error->log_exception($severity, $message, $filepath, $line);
	}
}

// --------------------------------------------------------------------

/**
 * Remove Invisible Characters
 *
 * This prevents sandwiching null characters
 * between ascii characters, like Java\0script.
 *
 * @access	public
 * @param	string
 * @return	string
 */
if ( ! function_exists('remove_invisible_characters'))
{
	function remove_invisible_characters($str, $url_encoded = TRUE)
	{
		$non_displayables = array();

		// every control character except newline (dec 10)
		// carriage return (dec 13), and horizontal tab (dec 09)

		if ($url_encoded)
		{
			$non_displayables[] = '/%0[0-8bcef]/';	// url encoded 00-08, 11, 12, 14, 15
			$non_displayables[] = '/%1[0-9a-f]/';	// url encoded 16-31
		}

		$non_displayables[] = '/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]+/S';	// 00-08, 11, 12, 14-31, 127

		do
		{
			$str = preg_replace($non_displayables, '', $str, -1, $count);
		}
		while ($count);

		return $str;
	}
}

// ------------------------------------------------------------------------

/**
* Returns HTML escaped variable
*
* @access	public
* @param	mixed
* @return	mixed
*/
if ( ! function_exists('html_escape'))
{
	function html_escape($var)
	{
		if (is_array($var))
		{
			return array_map('html_escape', $var);
		}
		else
		{
			return htmlspecialchars($var, ENT_QUOTES, config_item('charset'));
		}
	}
}

/* End of file Common.php */
/* Location: ./system/core/Common.php */
